monitoring/configure-metrics-tools.sh

```diff
@@ -0,0 +1,94 @@
+#!/bin/bash
+
+# Copyright 2024 The KubeStellar Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -x # echo so that users can understand what is happening
+set -e # exit on error
+
+ctx="kind-kubeflex"
+wds="wds1"
+
+if [ "$1" == "--space-name" ]; then
+    ctx="$2"
+    shift 2
+fi
+
+SCRIPT_DIR=$(cd -- "$(dirname -- "${BASH_SOURCE[0]}")" &> /dev/null && pwd)
+
+# set core context
+kubectl config use-context $ctx
+
+: --------------------------------------------------------------------
+: Configure kubestellar controller for prometheus scraping
+: --------------------------------------------------------------------
+: 1. configure the rbac for the kube-rbac-proxy 
+sed s/%WDS_NS%/$wds-system/g kube-rbac-proxy-config.yaml | kubectl apply -f -
+
+: 2. Create svc for KS ctl manager
+kubectl -n $wds-system apply -f ${SCRIPT_DIR}/ks-ctl-manager-svc.yaml
+
+: 3. Create service monitor for KS ctl manager
+sed s/%WDS_NS%/$wds-system/g sm-ks-controller-manager.yaml | kubectl -n ks-monitoring apply -f -
+
+
+: --------------------------------------------------------------------
+: Configure kubestellar controller for pyroscope scraping
+: --------------------------------------------------------------------
+: 1. Open the pprof port:
+kubectl -n $wds-system get deploy kubestellar-controller-manager -o yaml | yq '(del(.status) |.spec.template.spec.containers.[1].ports[0].name |= "http")' | yq '.spec.template.spec.containers.[1].ports[0].protocol |= "TCP"' | yq '.spec.template.spec.containers.[1].ports[0].containerPort |= 8082' | kubectl --context $ctx apply --namespace=$wds-system -f -
+
+: 2. Add labels for pyroscope scraping:
+kubectl -n $wds-system get deploy kubestellar-controller-manager -o yaml | yq '.spec.template.metadata.annotations."profiles.grafana.com/cpu.port" |= "8082" |  .spec.template.metadata.annotations."profiles.grafana.com/cpu.scrape"|= "true" | .spec.template.metadata.annotations."profiles.grafana.com/goroutine.port" |= "8082" | .spec.template.metadata.annotations."profiles.grafana.com/goroutine.scrape" |= "true" |
+.spec.template.metadata.annotations."profiles.grafana.com/memory.port" |= "8082" | .spec.template.metadata.annotations."profiles.grafana.com/memory.scrape" |= "true"' | kubectl -n $wds-system apply -f -
```

Comment:
See #2362

Reply:
Or, could this be indirected through a Service (as in Prometheus scraping)? That is, define a Service and put the needed labels/annotations on the Service?

Reply:
Pyroscope to scrape pods directly by adding annotations to the pods - not sure if it can scrape it through a service - I'll look into it.

Reply:
As I mentioned elsewhere, I have lately remembered that pulling from Pods is better than pulling from Services.